server/logger.go

```diff
@@ -0,0 +1,32 @@
+package server
+
+import (
+	"fmt"
+	"log"
+)
+
+type Logger interface {
+	log(format string, v ...interface{})
+	logError(format string, v ...interface{})
+}
+
+type Log struct {
+	uid, ip, origin string
+}
+
+func NewLogger(uid, ip, origin string) Logger {
+	return &Log{
+		uid:    uid,
+		ip:     ip,
+		origin: origin,
+	}
+}
+func (l *Log) log(format string, v ...interface{}) {
+	prefix := fmt.Sprintf("[%s] [%s] [%s] ", l.uid, l.ip, l.origin)
+	log.Printf(prefix+format, v...)
+}
+
+func (l *Log) logError(format string, v ...interface{}) {
+	prefix := fmt.Sprintf("[%s] [%s] [%s] ERROR: ", l.uid, l.ip, l.origin)
+	log.Printf(prefix+format, v...)
+}
```

Comment:
Perhaps we could add a function that returns a child logger where the original uid is appended with a suffix:

```
func (l *Log) CreateChildLogger(suffix string) {
```

This way we could pass a derived logger into each batched request to distinguish the log messages while still keeping a common base uid 🤔

Reply:
I have actually tried this yesteday But it will update the same underlying struct which will have problem while running go routine. One way we could solve this by creating new logger before starting the new go routine though It will create new object.
```
`go func(count int, rpcReq *types.JsonRpcRequest) {
			// Create logger and pass in go routine number
			l := NewLogger(r.uid + "-" + strconv.FormatInt(int64(count), 10))
			// Create rpc request
			req := NewRpcRequest(l, rpcReq, r.defaultProxyUrl, r.relaySigningKey, ip, origin) // Set each individual request
			res := req.ProcessRequest()
			resCh <- res
		}(i, jsonBatchReq[i])`
```

Reply:
```
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-3]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-35]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-6]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-0]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-7]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-1]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
2022/01/10 11:18:37 [82074df3-79b8-4ec3-861e-68c565d424d3-2]proxyRequest to: https://mainnet.infura.io/v3/77582faac1cc4bd7be282243cb13afa1
```

Reply:
What's the problem with the main logger creating children instances?

Reply:
```
2022/01/10 11:59:55 [a553c13e-5d2b-4249-ac35-ee26f9188107-0-1-35-17-13-10-9-2-11-14-12-26-5-27-18-28-19-29-3-30-4-31-22-32-33-34-24-23-25-7-6-8-20-21-16]Proxy to node successful: eth_getBlockByNumber
2022/01/10 11:59:55 [a553c13e-5d2b-4249-ac35-ee26f9188107-0-1-35-17-13-10-9-2-11-14-12-26-5-27-18-28-19-29-3-30-4-31-22-32-33-34-24-23-25-7-6-8-20-21-16]proxy response 200 after 0.641626 sec
2022/01/10 11:59:55 [a553c13e-5d2b-4249-ac35-ee26f9188107-0-1-35-17-13-10-9-2-11-14-12-26-5-27-18-28-19-29-3-30-4-31-22-32-33-34-24-23-25-7-6-8-20-21-16]proxy response 200 after 0.639939 sec
2022/01/10 11:59:55 [a553c13e-5d2b-4249-ac35-ee26f9188107-0-1-35-17-13-10-9-2-11-14-12-26-5-27-18-28-19-29-3-30-4-31-22-32-33-34-24-23-25-7-6-8-20-21-16]Proxy to node successful: eth_getBlockByNumber
2022/01/10 11:59:55 [a553c13e-5d2b-4249-ac35-ee26f9188107-0-1-35-17-13-10-9-2-11-14-12-26-5-27-18-28-19-29-3-30-4-31-22-32-33-34-24-23-25-7-6-8-20-21-16]Proxy to node successful: eth_getBlockByNumber
2022/01/10 11:59:55 [a553c13e-5d2b-4249-ac35-ee26f9188107-0-1-35-17-13-10-9-2-11-14-12-26-5-27-18-28-19-29-3-30-4-31-22-32-33-34-24-23-25-7-6-8-20-21-16]request took 0.643962 sec

```

Reply:
Now I got what you mean.like this ? 
```
func (l *Log) CreateChildLogger(suffix string) Logger {
	return NewLogger(l.uid + "-" + suffix)
}
```
This is same as creating a new logger right :-)

Reply:
Done